lib/core/base/virtual-node.js

```diff
@@ -0,0 +1,79 @@
+// class is unused in the file...
+// eslint-disable-next-line no-unused-vars
+class VirtualNode {
+	/**
+	 * Wrap the real node and provide list of the flattened children
+	 *
+	 * @param node {Node} - the node in question
+	 * @param shadowId {String} - the ID of the shadow DOM to which this node belongs
+	 */
+	constructor(node, shadowId) {
+		this.shadowId = shadowId;
+		this.children = [];
+		this.actualNode = node;
+
+		this._isHidden = null; // will be populated by axe.utils.isHidden
+		this._cache = {};
+
+		// abstract Node and Element APIs so we can run axe in DOM-less
+		// environments. these are static properties in the assumption
+		// that axe does not change any of them while it runs.
+		this.elementNodeType = node.nodeType;
+		this.elementNodeName = node.nodeName.toLowerCase();
+		this.elementId = node.id;
+
+		if (axe._cache.get('nodeMap')) {
+			axe._cache.get('nodeMap').set(node, this);
+		}
+	}
+
+	/**
+	 * Determine if the actualNode has the given class name.
+	 * @see https://j11y.io/jquery/#v=2.0.3&fn=jQuery.fn.hasClass
+	 * @param {String} className - The class to check for.
+	 * @return {Boolean|undefined} True if the actualNode has the given class, false if it does not have the given class, and undefined if the actualNode does not have a className property.
```

Comment:
Why return `undefined` rather than `false`? Having multiple return types (even tho documented) for a function prefixed with `has` seems like a very strange pattern.